beacon_node/beacon_chain/src/migrate.rs

```diff
@@ -79,75 +120,110 @@ pub trait Migrate<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>>:
         let mut abandoned_heads: HashSet<Hash256> = HashSet::new();
 
         for (head_hash, head_slot) in head_tracker.heads() {
-            let mut potentially_abandoned_head: Option<Hash256> = Some(head_hash);
-            let mut potentially_abandoned_blocks: Vec<(
-                Slot,
-                Option<SignedBeaconBlockHash>,
-                Option<BeaconStateHash>,
-            )> = Vec::new();
+            let mut potentially_abandoned_head = Some(head_hash);
+            let mut potentially_abandoned_blocks = vec![];
 
             let head_state_hash = store
                 .get_block(&head_hash)?
                 .ok_or_else(|| BeaconStateError::MissingBeaconBlock(head_hash.into()))?
                 .state_root();
 
+            // Iterate backwards from this head, staging blocks and states for deletion.
             let iter = std::iter::once(Ok((head_hash, head_state_hash, head_slot)))
-                .chain(RootsIterator::from_block(Arc::clone(&store), head_hash)?);
+                .chain(RootsIterator::from_block(store.clone(), head_hash)?);
```

Comment:
It seems to me that this iterator doesn't iterate across state roots from skipped slots. I would think we need to do a state root iter across `(Option<SignedBeaconBlockHash>, BeaconStateHash)` to clean up those skip-slot states. Perhaps I'm missing something though :)

Reply:
Nah the `RootsIterator` uses the block roots and state roots from a beacon state, so it returns something for every slot (including the roots of skipped states).

https://github.com/sigp/lighthouse/blob/175471a64b90fb368a27e3ac9377153cf9af7653/beacon_node/store/src/iter.rs#L190-L216